File: brixo-api/src/main/java/com/brixo/service/LlmService.java
package com.brixo.service;

import com.brixo.dto.CotizacionResult;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.*;

/**
 * Servicio multi-proveedor de LLM (Anthropic Claude, OpenAI, Groq).
 *
 * Replica la lógica de app/Libraries/LlmService.php del sistema PHP legacy.
 * Genera cotizaciones estructuradas en JSON a partir de descripciones de trabajo.
 */
@Service
public class LlmService {

    private static final Logger log = LoggerFactory.getLogger(LlmService.class);

    private final ObjectMapper mapper;

    @Value("${llm.provider:groq}")
    private String provider;

    @Value("${llm.api-key:}")
    private String apiKey;

    @Value("${llm.model:}")
    private String model;

    /** Prompt del sistema — fuerza respuesta JSON estructurada. */
    private static final String SYSTEM_PROMPT = """
            Eres un experto estimador de costos para servicios del hogar en Colombia.
            Respondes EXCLUSIVAMENTE en formato JSON válido, sin markdown, sin texto adicional.
            
            El JSON debe tener esta estructura exacta:
            {
              "servicio_principal": "nombre del servicio",
              "materiales": [
                {"nombre": "material", "cantidad": 1, "precio_unitario": 10000, "subtotal": 10000}
              ],
              "personal": [
                {"rol": "profesional", "cantidad": 1, "costo_hora": 25000, "horas": 8, "subtotal": 200000}
              ],
              "complejidad": "BAJA|MEDIA|ALTA",
              "total_estimado": 250000
            }
            
            Todos los precios en pesos colombianos (COP). Sé realista con los precios del mercado colombiano.
            """;

    public LlmService(ObjectMapper mapper) {
        this.mapper = mapper;
    }

    /**
     * Genera una cotización a partir de una descripción de trabajo.
     *
     * @param descripcion descripción libre del trabajo
     * @return resultado estructurado de la cotización
     */
    public Optional<CotizacionResult> generarCotizacion(String descripcion) {
        if (apiKey.isBlank()) {
            log.warn("LLM API key no configurada. Retornando cotización demo.");
            return Optional.of(demoCotizacion(descripcion));
        }

        try {
            String response = callLlm(descripcion);
            return parseResponse(response);
        } catch (Exception e) {
            log.error("Error al llamar LLM: {}", e.getMessage());
            return Optional.empty();
        }
    }

    private String callLlm(String userMessage) {
        return switch (provider.toLowerCase()) {
            case "anthropic", "claude" -> callAnthropic(userMessage);
            case "openai", "gpt"       -> callOpenAI(userMessage);
            case "groq"                -> callGroq(userMessage);
            default -> throw new IllegalArgumentException("Proveedor LLM no soportado: " + provider);
        };
    }

    // ═══════════════════════════════════════════
    // Anthropic Claude
    // ═══════════════════════════════════════════
    private String callAnthropic(String userMessage) {
        String effectiveModel = model.isBlank() ? "claude-sonnet-4-20250514" : model;

        Map<String, Object> body = Map.of(
                "model", effectiveModel,
                "max_tokens", 2048,
                "system", SYSTEM_PROMPT,
                "messages", List.of(Map.of("role", "user", "content", userMessage))
        );

        String response = WebClient.create("https://api.anthropic.com")
                .post()
                .uri("/v1/messages")
                .header("x-api-key", apiKey)
                .header("anthropic-version", "2023-06-01")
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        return extractAnthropicContent(response);
    }

    // ═══════════════════════════════════════════
    // OpenAI
    // ═══════════════════════════════════════════
    private String callOpenAI(String userMessage) {
        String effectiveModel = model.isBlank() ? "gpt-4o-mini" : model;

        Map<String, Object> body = Map.of(
                "model", effectiveModel,
                "messages", List.of(
                        Map.of("role", "system", "content", SYSTEM_PROMPT),
                        Map.of("role", "user", "content", userMessage)
                ),
                "temperature", 0.3
        );

        String response = WebClient.create("https://api.openai.com")
                .post()
                .uri("/v1/chat/completions")
                .header("Authorization", "Bearer " + apiKey)
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        return extractOpenAIContent(response);
    }

    // ═══════════════════════════════════════════
    // Groq
    // ═══════════════════════════════════════════
    private String callGroq(String userMessage) {
        String effectiveModel = model.isBlank() ? "llama-3.3-70b-versatile" : model;

        Map<String, Object> body = Map.of(
                "model", effectiveModel,
                "messages", List.of(
                        Map.of("role", "system", "content", SYSTEM_PROMPT),
                        Map.of("role", "user", "content", userMessage)
                ),
                "temperature", 0.3
        );

        String response = WebClient.create("https://api.groq.com")
                .post()
                .uri("/openai/v1/chat/completions")
                .header("Authorization", "Bearer " + apiKey)
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(String.class)
                .block();

        return extractOpenAIContent(response); // Groq usa el mismo formato que OpenAI
    }

    // ═══════════════════════════════════════════
    // Parseo de respuestas
    // ═══════════════════════════════════════════

    private String extractAnthropicContent(String response) {
        try {
            JsonNode root = mapper.readTree(response);
            return root.at("/content/0/text").asText();
        } catch (JsonProcessingException e) {
            log.error("Error parseando respuesta Anthropic: {}", e.getMessage());
            return response;
        }
    }

    private String extractOpenAIContent(String response) {
        try {
            JsonNode root = mapper.readTree(response);
            return root.at("/choices/0/message/content").asText();
        } catch (JsonProcessingException e) {
            log.error("Error parseando respuesta OpenAI/Groq: {}", e.getMessage());
            return response;
        }
    }

    private Optional<CotizacionResult> parseResponse(String raw) {
        try {
            // Limpiar posibles envolturas de markdown
            String clean = raw.replaceAll("```json\\s*", "").replaceAll("```\\s*", "").trim();

            JsonNode node = mapper.readTree(clean);

            List<CotizacionResult.MaterialItem> materiales = new ArrayList<>();
            for (JsonNode m : node.withArray("materiales")) {
                materiales.add(new CotizacionResult.MaterialItem(
                        m.get("nombre").asText(),
                        m.get("cantidad").asInt(),
                        m.get("precio_unitario").asDouble(),
                        m.get("subtotal").asDouble()
                ));
            }

            List<CotizacionResult.PersonalItem> personal = new ArrayList<>();
            for (JsonNode p : node.withArray("personal")) {
                personal.add(new CotizacionResult.PersonalItem(
                        p.get("rol").asText(),
                        p.get("cantidad").asInt(),
                        p.get("costo_hora").asDouble(),
                        p.get("horas").asInt(),
                        p.get("subtotal").asDouble()
                ));
            }

            return Optional.of(new CotizacionResult(
                    node.get("servicio_principal").asText(),
                    materiales,
                    personal,
                    node.get("complejidad").asText(),
                    node.get("total_estimado").asDouble()
            ));

        } catch (Exception e) {
            log.error("Error parseando respuesta LLM: {}", e.getMessage());
            return Optional.empty();
        }
    }

    /**
     * Cotización demo para cuando no hay API key configurada.
     */
    private CotizacionResult demoCotizacion(String descripcion) {
        return new CotizacionResult(
                "Servicio general del hogar",
                List.of(
                        new CotizacionResult.MaterialItem("Materiales básicos", 1, 50000, 50000),
                        new CotizacionResult.MaterialItem("Herramientas", 1, 30000, 30000)
                ),
                List.of(
                        new CotizacionResult.PersonalItem("Profesional", 1, 25000, 8, 200000)
                ),
                "MEDIA",
                280000
        );
    }
}

File: app/Libraries/LlmService.php
<?php

namespace App\Libraries;

/**
 * LlmService – Wrapper para APIs de LLM (Anthropic Claude / OpenAI / Groq).
 *
 * Lee las claves desde .env:
 *   LLM_PROVIDER   = anthropic | openai | groq
 *   LLM_API_KEY    = sk-... | gsk_...
 *   LLM_MODEL      = claude-sonnet-4-20250514 | gpt-4o-mini | llama-3.3-70b-versatile  (opcional)
 */
class LlmService
{
    /** @var string 'anthropic' | 'openai' | 'groq' */
    private string $provider;

    /** @var string API key */
    private string $apiKey;

    /** @var string Modelo a utilizar */
    private string $model;

    /** Prompt de sistema estricto para cotizaciones */
    private const SYSTEM_PROMPT = <<<'PROMPT'
Eres un asistente experto en construcción, remodelación y servicios del hogar.
Tu ÚNICA función es generar cotizaciones desglosadas a partir de la descripción
que proporciona el usuario.

REGLAS ESTRICTAS:
1. Responde ÚNICAMENTE con un objeto JSON válido. Sin texto antes ni después.
2. No uses bloques de código Markdown (```). Solo JSON puro.
3. El JSON DEBE ajustarse exactamente al siguiente esquema:

{
  "servicio_principal": "string – nombre corto del servicio",
  "materiales": [
    { "nombre": "string", "cantidad_estimada": "string con unidad" }
  ],
  "personal": [
    { "rol": "string – p.ej. Plomero, Albañil", "horas_estimadas": number }
  ],
  "complejidad": "bajo | medio | alto"
}

4. Si la descripción es ambigua, haz suposiciones razonables pero NO pidas más datos.
5. Incluye al menos 1 material y 1 rol de personal.
6. La complejidad debe ser exactamente uno de: "bajo", "medio" o "alto".
PROMPT;

    // ----------------------------------------------------------------
    // Constructor
    // ----------------------------------------------------------------

    public function __construct()
    {
        $this->provider = env('LLM_PROVIDER', 'groq');
        $this->apiKey   = env('LLM_API_KEY', '');

        $defaultModel = match ($this->provider) {
            'openai'    => 'gpt-4o-mini',
            'groq'      => 'llama-3.3-70b-versatile',
            default     => 'claude-sonnet-4-20250514',
        };
        $this->model = env('LLM_MODEL', $defaultModel);
    }

    // ----------------------------------------------------------------
    // Método público: genera cotización
    // ----------------------------------------------------------------

    /**
     * Envía la descripción del usuario al LLM y devuelve la cotización como array.
     *
     * @param  string $descripcionUsuario  Texto libre del cliente.
     * @return array{ok: bool, data?: array, error?: string}
     */
    public function generarCotizacion(string $descripcionUsuario): array
    {
        if (empty($this->apiKey)) {
            return $this->error('No se ha configurado LLM_API_KEY en .env');
        }

        $descripcionUsuario = trim($descripcionUsuario);
        if ($descripcionUsuario === '') {
            return $this->error('La descripción del servicio no puede estar vacía.');
        }

        // Llamar al proveedor adecuado
        $raw = match ($this->provider) {
            'openai'    => $this->callOpenAI($descripcionUsuario),
            'anthropic' => $this->callAnthropic($descripcionUsuario),
            'groq'      => $this->callGroq($descripcionUsuario),
            default     => null,
        };

        if ($raw === null) {
            return $this->error("Proveedor LLM no soportado: {$this->provider}");
        }

        if (isset($raw['_error'])) {
            return $this->error($raw['_error']);
        }

        // Extraer el texto de respuesta
        $content = $this->extractContent($raw);
        if ($content === null) {
            log_message('error', '[LlmService] Respuesta inesperada del LLM: ' . json_encode($raw));
            return $this->error('La IA devolvió una respuesta inesperada.');
        }

        // Limpiar posible bloque Markdown ```json ... ```
        $content = $this->limpiarMarkdown($content);

        // Decodificar JSON
        $cotizacion = json_decode($content, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            log_message('error', '[LlmService] JSON inválido del LLM: ' . $content);
            return $this->error('La IA no devolvió un JSON válido. Intenta reformular tu solicitud.');
        }

        // Validar esquema mínimo
        $validacion = $this->validarEsquema($cotizacion);
        if ($validacion !== true) {
            log_message('error', '[LlmService] Esquema inválido: ' . $validacion);
            return $this->error("Formato de respuesta incorrecto: {$validacion}");
        }

        return ['ok' => true, 'data' => $cotizacion];
    }

    // ----------------------------------------------------------------
    // Llamadas a APIs
    // ----------------------------------------------------------------

    /**
     * Llama a la API de Anthropic (Messages API v1).
     */
    private function callAnthropic(string $userMessage): ?array
    {
        $url = 'https://api.anthropic.com/v1/messages';

        $payload = [
            'model'      => $this->model,
            'max_tokens' => 1024,
            'system'     => self::SYSTEM_PROMPT,
            'messages'   => [
                ['role' => 'user', 'content' => $userMessage],
            ],
        ];

        $headers = [
            'Content-Type: application/json',
            'x-api-key: ' . $this->apiKey,
            'anthropic-version: 2023-06-01',
        ];

        return $this->curlPost($url, $payload, $headers);
    }

    /**
     * Llama a la API de OpenAI (Chat Completions).
     */
    private function callOpenAI(string $userMessage): ?array
    {
        $url = 'https://api.openai.com/v1/chat/completions';

        $payload = [
            'model'       => $this->model,
            'max_tokens'  => 1024,
            'temperature' => 0.3,
            'messages'    => [
                ['role' => 'system',  'content' => self::SYSTEM_PROMPT],
                ['role' => 'user',    'content' => $userMessage],
            ],
        ];

        $headers = [
            'Content-Type: application/json',
            'Authorization: Bearer ' . $this->apiKey,
        ];

        return $this->curlPost($url, $payload, $headers);
    }

    /**
     * Llama a la API de Groq (compatible con formato OpenAI).
     * Usa response_format json_object para garantizar JSON limpio.
     */
    private function callGroq(string $userMessage): ?array
    {
        $url = 'https://api.groq.com/openai/v1/chat/completions';

        $payload = [
            'model'           => $this->model,
            'max_tokens'      => 1024,
            'temperature'     => 0.3,
            'response_format' => ['type' => 'json_object'],
            'messages'        => [
                ['role' => 'system',  'content' => self::SYSTEM_PROMPT],
                ['role' => 'user',    'content' => $userMessage],
            ],
        ];

        $headers = [
            'Content-Type: application/json',
            'Authorization: Bearer ' . $this->apiKey,
        ];

        return $this->curlPost($url, $payload, $headers);
    }

    // ----------------------------------------------------------------
    // Utilidades internas
    // ----------------------------------------------------------------

    /**
     * Ejecuta un POST con cURL y devuelve el body como array.
     */
    private function curlPost(string $url, array $payload, array $headers): array
    {
        $ch = curl_init($url);

        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => $headers,
            CURLOPT_POSTFIELDS     => json_encode($payload),
            CURLOPT_TIMEOUT        => 30,
            CURLOPT_CONNECTTIMEOUT => 10,
        ]);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $curlError = curl_error($ch);
        curl_close($ch);

        if ($response === false) {
            return ['_error' => "Error de conexión con la API: {$curlError}"];
        }

        $decoded = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            return ['_error' => "Respuesta no-JSON de la API (HTTP {$httpCode})."];
        }

        // Manejar errores HTTP del proveedor
        if ($httpCode >= 400) {
            $msg = $decoded['error']['message'] ?? ($decoded['error']['type'] ?? "HTTP {$httpCode}");
            return ['_error' => "Error de la API ({$httpCode}): {$msg}"];
        }

        return $decoded;
    }

    /**
     * Extrae el texto de contenido según el formato de cada proveedor.
     */
    private function extractContent(array $response): ?string
    {
        // Anthropic: { content: [{ type: "text", text: "..." }] }
        if (isset($response['content'][0]['text'])) {
            return $response['content'][0]['text'];
        }

        // OpenAI / Groq: { choices: [{ message: { content: "..." } }] }
        if (isset($response['choices'][0]['message']['content'])) {
            return $response['choices'][0]['message']['content'];
        }

        return null;
    }

    /**
     * Limpia bloques de código Markdown que algunos modelos incluyen.
     */
    private function limpiarMarkdown(string $text): string
    {
        // Quitar ```json ... ``` o ``` ... ```
        $text = preg_replace('/^```(?:json)?\s*/i', '', trim($text));
        $text = preg_replace('/\s*```$/i', '', $text);
        return trim($text);
    }

    /**
     * Valida que el array tenga el esquema esperado.
     *
     * @return true|string  true si es válido, o string con mensaje de error.
     */
    private function validarEsquema(array $data): true|string
    {
        if (empty($data['servicio_principal']) || !is_string($data['servicio_principal'])) {
            return 'Falta "servicio_principal" (string).';
        }

        if (empty($data['materiales']) || !is_array($data['materiales'])) {
            return 'Falta "materiales" (array).';
        }
        foreach ($data['materiales'] as $i => $mat) {
            if (empty($mat['nombre'])) {
                return "materiales[{$i}] no tiene \"nombre\".";
            }
            if (!isset($mat['cantidad_estimada'])) {
                return "materiales[{$i}] no tiene \"cantidad_estimada\".";
            }
        }

        if (empty($data['personal']) || !is_array($data['personal'])) {
            return 'Falta "personal" (array).';
        }
        foreach ($data['personal'] as $i => $per) {
            if (empty($per['rol'])) {
                return "personal[{$i}] no tiene \"rol\".";
            }
            if (!isset($per['horas_estimadas'])) {
                return "personal[{$i}] no tiene \"horas_estimadas\".";
            }
        }

        $validas = ['bajo', 'medio', 'alto'];
        if (empty($data['complejidad']) || !in_array($data['complejidad'], $validas, true)) {
            return '"complejidad" debe ser "bajo", "medio" o "alto".';
        }

        return true;
    }

    /**
     * Genera un array de error estandarizado.
     */
    private function error(string $msg): array
    {
        return ['ok' => false, 'error' => $msg];
    }
}
